test/testutil.h

```diff
@@ -141,10 +143,94 @@ __owur int run_tests(const char *test_prog_name);
+ *
+ * |desc| is a printf formatted description with arguments |args|, |type| is
+ * the data type that was tested, |fmt| is a system provided printf format with
+ * arguments that spells out the failure details.
+ */
+#if defined(__GNUC__)
+void fail_message(const char *desc, va_list args, const char *type,
+                  const char *fmt, ...) __attribute__ ((format(printf, 4, 5)));
+
+#define DECLARE_COMPARISON(type, name, opname)                          \
+    int name ## _ ## opname(const type, const type, const char *, ...)  \
+            __attribute__ ((format(printf, 3, 4)));
+#else
+void fail_message(const char *desc, va_list args, const char *type,
+                         const char *fmt, ...);
+
+#define DECLARE_COMPARISON(type, name, opname)                          \
+    int name ## _ ## opname(const type, const type,  const char *, ...)
+#endif
+
+#define DECLARE_COMPARISONS(type, name)                                 \
+    DECLARE_COMPARISON(type, name, eq)                                  \
+    DECLARE_COMPARISON(type, name, ne)                                  \
+    DECLARE_COMPARISON(type, name, lt)                                  \
+    DECLARE_COMPARISON(type, name, le)                                  \
+    DECLARE_COMPARISON(type, name, gt)                                  \
+    DECLARE_COMPARISON(type, name, ge)
+
+#define DEFINE_COMPARISON(type, name, opname, op, fmt)                  \
+    int name ## _ ## opname(const type t1, const type t2,               \
+                            const char *desc, ...)                      \
+    {                                                                   \
+        va_list ap;                                                     \
+        if (t1 op t2)                                                   \
+            return 1;                                                   \
+        va_start(ap, desc);                                             \
+        fail_message(desc, ap, #type, fmt " " #opname " " fmt, t1, t2); \
+        va_end(ap);                                                     \
+        return 0;                                                       \
+    }
+
+#define DEFINE_COMPARISONS(type, name, fmt)                             \
+    DEFINE_COMPARISON(type, name, eq, ==, fmt)                          \
+    DEFINE_COMPARISON(type, name, ne, !=, fmt)                          \
+    DEFINE_COMPARISON(type, name, lt, <, fmt)                           \
+    DEFINE_COMPARISON(type, name, le, <=, fmt)                          \
+    DEFINE_COMPARISON(type, name, gt, >, fmt)                           \
+    DEFINE_COMPARISON(type, name, ge, >=, fmt)
+
+#define IMPLEMENT_COMPARISONS(type, name, fmt)                          \
+    DECLARE_COMPARISONS(type, name)                                     \
+    DEFINE_COMPARISONS(type, name, fmt)
+
+/*
+ * Define eight comparisons between pairs of integers and unsigned integers.
```

Comment:
It'd be nice if you could spell out, at least once here, what the resulting method signatures are.